### ISSUE TITLE
Training Deep Neural Network

### ISSUE THREAD

**Issue by rajareddy1863** [fchollet/keras#1013]
I am trying to train a deep neural network for speech recognition using MFCC coefficients. I have 2000 hidden units per layer and a total of 6 layers. Even then my network does not seem to learn anything. Even though the value of loss function decreases accuracy on the training set does not seem to improve. Any suggestions?

Thanks in advance.

**Comment by xtinkrr** [comment]
I'm not quite sure what your problem is. Perhaps post the part of your code where you build the model? I would guess, either you are evaluating the loss function improperly, or the weight decay regularization parameter/s (if you are using one) is/are too large, or you have far too few (training/test) samples for your model.

**Comment by rajareddy1863** [comment]
Thanks for your reply. I am pasting the code here. I would be very much grateful if you can provide me a way out. 

Npte: I have training data of around 9 lakh samples 

n_output = 183
n_int = 2000
model = Sequential()
model.add(Dense(n_int, input_dim = 60, init = 'uniform'))
model.add(Activation('relu'))
#model.add(Dropout(0.2))
model.add(Dense(n_int, init = 'uniform'))
model.add(Activation('relu'))
model.add(Dense(n_int, init = 'uniform'))
model.add(Activation('relu'))
model.add(Dropout(0.2))
model.add(Dense(n_int, init = 'uniform'))
model.add(Activation('relu'))
model.add(Dropout(0.2))
model.add(Dense(n_int, init = 'uniform'))
model.add(Activation('relu'))
model.add(Dropout(0.2))
model.add(Dense(n_int, init = 'uniform'))
model.add(Activation('relu'))
model.add(Dropout(0.2))
model.add(Dense(n_int, init = 'uniform'))
model.add(Activation('relu'))
model.add(Dropout(0.2))
model.add(Dense(n_output, init = 'uniform'))
model.add(Activation('relu'))
model.add(Dropout(0.2))
sgd = SGD(lr = 0.01, decay = 1e-6, momemtum = 0.9, nesterov = True)
model.compile(loss='categorical_crossentropy', optimizer=sgd)

model.fit(trainSet, labelSet, nb_epoch = 3, batch_size = 512, show_accuracy = True, verbose = 1)

**Comment by xtinkrr** [comment]
Oh, so I see you are doing a 183-class classification problem. Then I suggest you should change your last layer into a softmax layer instead of a Relu with Dropout. The softmax function transforms your hidden units into probability scores of the class labels you have. See the wiki entry for more info: https://en.wikipedia.org/wiki/Softmax_function.

So basically, this edit should work:
[0] model.add(Dense(n_output, init = 'uniform'))
[-] model.add(Activation('relu'))
[-] model.add(Dropout(0.2))
[+] model.add(Activation('softmax'))
[0] sgd = SGD(lr = 0.01, decay = 1e-6, momemtum = 0.9, nesterov = True)

[-] = delete, [+] = add, [0] = no_change

You don't need Dropout in your last layer.

Another tip (but perhaps irrelevant), try starting with a simpler model first. Maybe a couple (2 or 3) of hidden layers and fewer hidden units (~100-300). Because they are way easier and faster to train. And just move to more complicated, if you need to. For example, if you observe underfitting (low training and test accuracy, but both are relatively close to each other).

P.S. I don't know how many 9 lakh are. :P
I need to ask my Indian friends. I hope my post helps.

**Comment by rajareddy1863** [comment]
First of all thank you very much for your early reply. Changing the last layer's activation function to 'softmax' worked. I am now trying to work with 20000 samples (60 dimensions each) with 300 hidden units and 4 hidden layers. I am not considering any dropout. But now the accuracy seems to saturate after some time. Do you have any idea on why this happens. I am would be much grateful if you can provide me a way to choose number of hidden units, hidden layers depending on the training data. 

Neural Network....
Epoch 1/7
21s - loss: 5.1384 - acc: 0.0602
Epoch 2/7
21s - loss: 5.0107 - acc: 0.0605
Epoch 3/7
21s - loss: 4.9075 - acc: 0.0605
Epoch 4/7
21s - loss: 4.8279 - acc: 0.0605
Epoch 5/7
21s - loss: 4.7673 - acc: 0.0605
Epoch 6/7
21s - loss: 4.7205 - acc: 0.0605
Epoch 7/7
21s - loss: 4.6831 - acc: 0.0605

Thanks in advance.

**Comment by xtinkrr** [comment]
From what I see in your post, it's the same problem as your original post. The loss function decreases but the training accuracy stops. I dunno if it has something to do with your problem using a large number of classes (have never worked on something like this before). [This](http://cs231n.github.io/neural-networks-2/#losses) tutorial suggest a "hierarchical softmax" loss function for large mult-class problems. Though I don't have enough knowledge how that works. Are you encoding, e.g. [one-hot encoding](https://en.wikipedia.org/wiki/One-hot) for hard classification problems, the training labels properly?

**Comment by xtinkrr** [comment]
And regarding the choice of number of hidden layers, hidden units, etc. It's another problem in statistical machine learning we call [model selection](https://en.wikipedia.org/wiki/Model_selection). To perform this, we usually apply one of the [cross validation](https://en.wikipedia.org/wiki/Cross-validation_(statistics)) methods such as [holdout](https://en.wikipedia.org/wiki/Cross-validation_(statistics)#2-fold_cross-validation).

**Comment by rajareddy1863** [comment]
Thanks for your early reply. I don't think that I have too many classes (I have only 183). Yes, I am encoding the training data. I have a number (0 to 182) assigned to each training sample (60 dimensional).

**Comment by xtinkrr** [comment]
If you are using the categorical loss function for classifying hard labels (only assign one class to each samples), you need to encode the labels using a method like [one-hot encoding](https://en.wikipedia.org/wiki/One-hot). Where for example you have labels: A, B, and C. You have to encode them into one-hot vectors: [1, 0, 0], [0, 1, 0], [0, 0, 1].

In keras, you can do it using:
```python
from keras.utils import np_utils
# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, nb_classes)
Y_test = np_utils.to_categorical(y_test, nb_classes)
```

**Comment by rajareddy1863** [comment]
Yes I was using that.

**Comment by xtinkrr** [comment]
As I can see from the training history you posted, the loss continues to goes down but the accuracy doesn't go up. What I can imagine is the probability score for the correctly classified samples increases (and thus the loss function decreases) but there are no (or at lea st insignificant) increase in the number of correctly classified samples.

Then I don't have anything new up my sleeve. I would say the same thing as I've said. If there is underfitting, you might want to increase the number of hidden layers/units. Maybe someone with more experience can help.

**Comment by udani969** [comment]
HI, I am also facing this situation. I have been trying to implement a classifier for numbers. I have 11 classes (10 for numbers and 1 for dummy label) I am also implementing this to classify speech data, but with a LSTM layer and a softmax layer. There are categorical labels asigned for frame per file. Like it is mentioned here, my network also reaches accuracy of 0.9899 at first epoch and then it saturates. But the loss keep on decaying with the rest of the epoches. I am also having the doubt about whether everything is working as expected.

**Comment by ersinyar** [comment]
@udani969 Do you mean train_acc of 0.9899 at first epoch? That seems odd. Because most of the time it starts from a lower value and goes up to this level at higher number of epochs. Did you manage to solve the problem? 
I am experiencing a similar situation for val_acc. It starts from 0.8 at first epoch and over epochs it does not change much. It only fluctuates between 0.8 and 0.83. Maybe your solution could be helpful for me.
